homeassistant/components/konnected/__init__.py

```diff
@@ -31,31 +37,169 @@
 from homeassistant.helpers import config_validation as cv
 
 from .config_flow import (  # Loading the config flow file will register the flow
-    DEVICE_SCHEMA_YAML,
+    CONF_DEFAULT_OPTIONS,
+    CONF_IO,
+    CONF_IO_BIN,
+    CONF_IO_DIG,
+    CONF_IO_SWI,
+    OPTIONS_SCHEMA,
 )
 from .const import (
     CONF_ACTIVATION,
     CONF_API_HOST,
+    CONF_BLINK,
+    CONF_DISCOVERY,
+    CONF_INVERSE,
+    CONF_MOMENTARY,
+    CONF_PAUSE,
+    CONF_POLL_INTERVAL,
+    CONF_REPEAT,
     DOMAIN,
     PIN_TO_ZONE,
     STATE_HIGH,
+    STATE_LOW,
     UPDATE_ENDPOINT,
     ZONE_TO_PIN,
+    ZONES,
 )
 from .errors import CannotConnect
 from .handlers import HANDLERS
 from .panel import AlarmPanel
 
 _LOGGER = logging.getLogger(__name__)
 
+
+def ensure_pin(value):
+    """Check if valid pin and coerce to string."""
+    if value is None:
+        raise vol.Invalid("pin value is None")
+
+    if PIN_TO_ZONE.get(str(value)) is None:
+        raise vol.Invalid("pin not valid")
+
+    return str(value)
+
+
+def ensure_zone(value):
+    """Check if valid zone and coerce to string."""
+    if value is None:
+        raise vol.Invalid("zone value is None")
+
+    if str(value) not in ZONES is None:
+        raise vol.Invalid("zone not valid")
+
+    return str(value)
+
+
+def import_validator(config):
+    """Validate zones and reformat for import."""
+    config = copy.deepcopy(config)
+    io_cfgs = {}
+    # Replace pins with zones
+    for zone in config.get(CONF_BINARY_SENSORS, []):
+        if zone.get(CONF_PIN):
+            zone[CONF_ZONE] = PIN_TO_ZONE[zone[CONF_PIN]]
+            del zone[CONF_PIN]
+        io_cfgs[zone[CONF_ZONE]] = CONF_IO_BIN
+    for zone in config.get(CONF_SENSORS, []):
+        if zone.get(CONF_PIN):
+            zone[CONF_ZONE] = PIN_TO_ZONE[zone[CONF_PIN]]
+            del zone[CONF_PIN]
+        io_cfgs[zone[CONF_ZONE]] = CONF_IO_DIG
+    for zone in config.get(CONF_SWITCHES, []):
+        if zone.get(CONF_PIN):
+            zone[CONF_ZONE] = PIN_TO_ZONE[zone[CONF_PIN]]
+            del zone[CONF_PIN]
+        io_cfgs[zone[CONF_ZONE]] = CONF_IO_SWI
+
+    # Migrate config_entry data into default_options structure
+    config[CONF_IO] = io_cfgs
+    config[CONF_DEFAULT_OPTIONS] = OPTIONS_SCHEMA(config)
+
+    # clean up fields migrated to options
+    config.pop(CONF_BINARY_SENSORS, None)
+    config.pop(CONF_SENSORS, None)
+    config.pop(CONF_SWITCHES, None)
+    config.pop(CONF_BLINK, None)
+    config.pop(CONF_DISCOVERY, None)
+    config.pop(CONF_IO, None)
+    return config
+
+
+# configuration.yaml schemas (legacy)
+BINARY_SENSOR_SCHEMA_YAML = vol.All(
+    vol.Schema(
+        {
+            vol.Exclusive(CONF_ZONE, "s_zone"): ensure_zone,
+            vol.Exclusive(CONF_PIN, "s_pin"): ensure_pin,
```

Comment:
Should `CONF_ZONE` and `CONF_PIN` be part of the same exclusion group?

Reply:
After the conversion to zone they should be but at this point in the validation they have some overlapping ID's.

Reply:
Without a companion in the same exclusion group, the exclusive mark becomes useless. What am I missing?

Reply:
You're correct - I was misinterpreting the scope of the exclusion.  Will fix